### PR TITLE
Add support for embedded iframe

### PR DESCRIPTION
**Pull request by mathiash98**
Goal: Add support for embedding IFrame in articles in order to render the interactive map.

Added content model in Contentful for IFrame:
image

To use:
1. Add new content with contentType `IFrame`
image
2. Edit article and click "Embed" -> "Entry" -> Select an IFrame entry
image

Demo of embedded map:
https://user-images.githubusercontent.com/5204006/174398406-4bbe1d0d-5876-4ab4-a931-1e2175162414.mp4

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,31 +1,23 @@
-# create-svelte
+# Bergen Triathlon website
+Official website for the Bergen Triathlon.
 
-Everything you need to build a Svelte project, powered by [`create-svelte`](https://github.com/sveltejs/kit/tree/master/packages/create-svelte);
+Frontend is written in Svelte with content being served from a ContentFul backend, currently at https://app.contentful.com/spaces/wi3d49ltoppx
```
Comment (jonalu) [inline README.md:4]: Please avoid publishing the Contentful space ID.

File: README.md [hunk 2]
```diff
@@ -1,31 +1,23 @@
...
-# create a new project in the current directory
-npm init svelte@next
+- [Node.js LTS](https://nodejs.org/en/)
 
-# create a new project in my-app
-npm init svelte@next my-app
+1. Add a file with name `.env` to add contentful tokens. Find the access token under `Settings` -> `API Keys` in contentful dashboard.
+```
+VITE_CONTENTFUL_DELIVERY_API_ACCESS_TOKEN=Fetch-Token-From-Contentful-backend
+VITE_CONTENTFUL_SPACE_ID=wi3d49ltoppx
```
Comment (jonalu) [inline README.md:15]: Omit the space ID. You could use a placeholder instead.

File: src/lib/contentful/types/Article.ts [hunk 3]
```diff
@@ -1,16 +1,48 @@
...
+	body: {
+		json: Document;
+		links: ArticleLinks;
 	};
-	articleHeroImage: {
+	articleHeroImage?: {
 		title: string;
 		url: string;
-	};
+	} | null;
```
Comment (jonalu) [inline src/lib/contentful/types/Article.ts:14]: The ` | null ` here seems superfluous as you've configured as `articleHeroImage? : { }`

File: src/routes/news/[slug].svelte [hunk 4]
```diff
@@ -67,8 +101,8 @@
 
         <img
                 class="rounded-xl mb-6 w-full"
-                alt={article.articleHeroImage.title}
-                src={article.articleHeroImage.url}
+                alt={article.articleHeroImage?.title}
+                src={article.articleHeroImage?.url}
         />
```
Comment (jonalu) [inline src/routes/news/[slug].svelte:106]: It is best not to render the image element altogether if the `url` property is `undefined`.

File: src/routes/news/[slug].svelte [hunk 5]
```diff
@@ -44,7 +54,31 @@
...
+                                width="${(entry as IFrameEntry).width}"
+                                frameBorder="0"
+                                scrolling="no"
+                                title="${(entry as IFrameEntry).title}"
+                            >Det ser ut til den innebygde iframen ikke ble lastet inn, prøv igjen senere eller kontakt sideadministrator.</iframe>
+                        `;
+                        return iframeElementString;
+                    } else {
+                        console.debug("Nothing");
+                        return "nothing";
```
Comment (jonalu) [inline src/routes/news/[slug].svelte:79]: Remove all debug logging. Also returning the string `"nothing"` makes no sense. It is better to return empty.

File: src/routes/news/[slug].svelte [hunk 6]
```diff
@@ -44,7 +54,31 @@
...
+                    if (entry.__typename === "IFrame") {
+                        const iframeElementString = `
+                            <iframe
+                                src="${(entry as IFrameEntry).url}"
+                                height="${(entry as IFrameEntry).height}"
+                                width="${(entry as IFrameEntry).width}"
+                                frameBorder="0"
+                                scrolling="no"
+                                title="${(entry as IFrameEntry).title}"
+                            >Det ser ut til den innebygde iframen ikke ble lastet inn, prøv igjen senere eller kontakt sideadministrator.</iframe>
```
Comment (jonalu) [inline src/routes/news/[slug].svelte:74]: Omit the fallback text. It does not give the user any valuable information. Also, there is no way for the user to contact the site administrator.

File: src/routes/news/[slug].svelte [hunk 7]
```diff
@@ -44,7 +54,31 @@
...
+                    // referenced in the GraphQL query
+                    if (entry.__typename === "IFrame") {
+                        const iframeElementString = `
+                            <iframe
+                                src="${(entry as IFrameEntry).url}"
+                                height="${(entry as IFrameEntry).height}"
+                                width="${(entry as IFrameEntry).width}"
+                                frameBorder="0"
+                                scrolling="no"
+                                title="${(entry as IFrameEntry).title}"
```
Comment (jonalu) [inline src/routes/news/[slug].svelte:73]: Try to follow the DRY principle and avoid casting the `entry` to `IFrameEntry` more than once.

File: src/routes/news/[slug].svelte [hunk 8]
```diff
@@ -44,7 +54,31 @@
                 [BLOCKS.EMBEDDED_ASSET]: (node, next) => {
                     const asset = assetMap.get(node.data.target.sys.id);
                     return `<img src=${asset.url} alt="${asset.title}" />`;
-                }
+                },
+                [BLOCKS.EMBEDDED_ENTRY]: (node, next) => {
+                    // https://www.contentful.com/blog/2021/04/14/rendering-linked-assets-entries-in-contentful/
```
Comment (jonalu) [inline src/routes/news/[slug].svelte:59]: This comment provides some value as it refers to relevant documentation. All other comments should be removed. We should strive to keep the code self documented through proper naming conventions.

File: src/lib/news.svelte [hunk 9]
```diff
@@ -9,10 +9,10 @@
 <div class="mx-auto grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-8 px-6">
 	{#each articles as article}
 		<ArticleLink
-			imageAltText={article.articleHeroImage.title}
+			imageAltText={article.articleHeroImage?.title}
 			url={`/news/${article.slug}`}
 			date={format(parseISO(article.date), 'MMM d, yy', { locale: nb })}
-			imageSrc={article.articleHeroImage.url}
+			imageSrc={article.articleHeroImage?.url}
```
Comment (jonalu) [inline src/lib/news.svelte:15]: How would the `<ArticleLink />` component handle missing `imageSrc`. Now that you've made this field optional, this scenario needs to be handled.

**Review by jonalu — CHANGES_REQUESTED** [review]
Thanks for your contribution. I have provided some feedback in the comments, please address these. As we haven't provided a style guide it is understandable that some code styles deviates a bit from the our standard.

File: src/lib/contentful/types/Article.ts (hunk 3 above)
Comment (jonalu) [reply src/lib/contentful/types/Article.ts:14]: Ok, remove the `?`

**Review by jonalu — COMMENTED** [review]
I would like to get your take on how you plan on making the iframe work in a responsive layout.

File: src/routes/news/[slug].svelte [hunk 10]
```diff
@@ -1,82 +1,113 @@
...
+					// https://www.contentful.com/blog/2021/04/14/rendering-linked-assets-entries-in-contentful/
+					const entry = entryMap.get(node.data.target.sys.id);
+
+					if (entry.__typename === 'IFrame') {
+						const iframeEntry = entry as IFrameEntry;
+						const iframeElementString = `
+                            <iframe
+                                src="${iframeEntry.url}"
+                                height="${iframeEntry.height}"
+                                width="${iframeEntry.width}"
```
[the remark this answers is not in the document]
Comment (mathiash98) [reply src/routes/news/[slug].svelte:74]: That method will not create a general responsive iframe as you assume the aspect ratio of the content. The map we are embedding is a website which is responsive by itself.

In general it's always hard to insert content that catches touch input into websites as the user can be stuck in that area. 
- So the important thing is to keep some margin on the sides to enable scrolling if the height of the screen is smaller than the height of the content.

Screenshot is of iPhone SE:
image
iPhone 12 pro:
image

**Comment by mathiash98** [comment]
> I have approved the Netlify preview release so that we can see the changes live.

Connected to another backend?

**Comment by jonalu** [comment]
> > I have approved the Netlify preview release so that we can see the changes live.
> 
> Connected to another contentful CMS backend?

Good question. Actually no. I have added support for Contenful draft content in Netlify preview-deploys:
https://github.com/Sonat-Consulting/b3/pull/7

If you update your PR with the `main` branch you should be able to add draft content in Contenful and view them in the Netlify preview deployment as seen [here](https://deploy-preview-7--bergen3.netlify.app/).

**Comment by jonalu** [comment]
Super. It looks good. For the future, try to avoid merge commits and rather rebase your changes on top of the `main` branch. That allows us to rebase the PRs.

**Pull request merged**